dynamiqs/time_tensor.py

```diff
@@ -239,3 +278,103 @@ def __add__(self, other: Tensor | TimeTensor) -> TimeTensor:
             return CallableTimeTensor(f, f0)
         else:
             return NotImplemented
+
+
+class _PWCTensor:
+    # Defined by a tuple of 2 tensors (times, values), where
+    # - times: (nv+1) are the time points between which the PWC tensor take constant
+    #          values, where nv is the number of time intervals
+    # - values: (..., nv) are the constant values for each time interval, where
+    #           (...) is an arbitrary batching size
+
+    def __init__(self, times: Tensor, values: Tensor):
+        self.times = times  # (nv+1)
+        self.values = values  # (..., nv)
+        self.nv = len(self.values)
+
+    @property
+    def shape(self) -> torch.Size:
+        return self.values.shape[:-1]  # (...)
+
+    def conj(self) -> _PWCTensor:
+        return _PWCTensor(self.times, self.values.conj())
+
+    def __call__(self, t: float) -> Tensor:
+        if t < self.times[0] or t >= self.times[-1]:
+            return torch.zeros_like(self.values[..., 0])  # (...)
+        else:
+            # find the index $k$ such that $t \in [t_k, t_{k+1})$
+            idx = torch.searchsorted(self.times, t, side='right') - 1
+            return self.values[..., idx]  # (...)
+
+    def view(self, *shape: int) -> _PWCTensor:
+        return _PWCTensor(self.times, self.values.view(*shape, self.nv))
+
+
+class PWCTimeTensor(TimeTensor):
+    # Arbitrary sum of tensors with PWC factors.
+
+    def __init__(
+        self, factors: list[_PWCTensor], tensors: Tensor, static: Tensor | None = None
+    ):
+        # factors must be non-empty
+        self.factors = factors  # list of length (nf)
+        self.tensors = tensors  # (nf, n, n)
+        self.n = tensors.shape[-1]
+        self.static = torch.zeros_like(self.tensors[0]) if static is None else static
+
+        # merge all times
+        self.times = torch.cat([x.times for x in self.factors]).unique(sorted=True)
+
+    @property
+    def dtype(self) -> torch.dtype:
+        return self.tensors.dtype
+
+    @property
+    def device(self) -> torch.device:
+        return self.tensors.device
+
+    @property
+    def shape(self) -> torch.Size:
+        return torch.Size((*self.factors[0].shape, self.n, self.n))  # (..., n, n)
+
+    def __call__(self, t: float) -> Tensor:
+        static = self.static.expand(*self.shape)  # (..., n, n)
```

Comment:
What is the point of this line? I think the sum should still work with broadcasting. Maybe it should be inside the first `if` statement.